### ISSUE TITLE
Certain combinations of check periods and check intervals prevent service status updates.

### ISSUE THREAD

**Issue by yoshi314** [Icinga/icinga2#9984]
## To Reproduce

I had a service that would check ssl cert validity. Mistakenly i configured it to have 

```
check_interval = 24h
check_period = "workhours"
```

which meant it would execute 7-21 on business days.

This check was set up to go critical if certificate was 7 days away from expiration. 

For a while this worked fine. but i've noticed that my check was stuck in warning state, saying i have 23 days left until Feb 2 - which (as of yesterday) was obviously incorrect. And the certificate was expiring in 2 days. So i had a stale service, except it seemed that it was being checked on schedule every 24 hours.

Upon inspection the check scheduled itself to run at ~4am, and apparently stopped updating service state for a few weeks since it was outside of its timeperiod. Manual reschedule of the check fixed the issue temporarily. 

It appeared as if the check was running on schedule, just not updating service state.

I had to readjust the check to have notification timeperiod assigned instead.

## Expected behavior

I assumed that icinga2 would try to schedule the check within the timeperiod window - especially if it was running late. Or would update the service state on the beginning of check_timeperiod.

I've seen this on 2.14.1 and 2.14.2 on Debian 10 (packages from icinga repository).

Maybe some kind of warning at config reload would suffice if check_interval may cause the service to "fall out"  of its check_period window.

**Comment by Al2Klimov** [comment]
> The scheduler should have the check_period in mind and should plan only checks when they are inside the next timeperiod

@julianbrost This would also solve our little overdue problem.

**Comment by Al2Klimov** [comment]
> Manual reschedule of the check fixed the issue temporarily.

This is indeed a legit workaround. You could also make the check interval just 1h lower than your check window, 13h I guess.

**Comment by SebastianOpeni** [comment]
We also observed this behavior and would like to add a follow-up problem.

tl/dr:
It can happen that the check_interval schedules the check at the end of a once-per-day check_period and then the retry_interval never schedules a retry until the next day.

Problem:
We have a (resourcewise) expensive check that sometimes fails on the first run.
Having experienced the discussed issue in the past, we configured the check_period and the check_interval to the same length. (15m)
Now we get exactly one check every day (with a slightly confusing view in icingadb-web).

But it is still no perfect solution since the schedule of the retry_interval is also blocked by the check_period.
This is totally expected and good, but because of that we have run out of workarounds  :see_no_evil:

One possible fix would be the above discussed reschedule with regards to the period. (preferred)
But if that is not possible, an idea could also be to offer a retry_period property that can then be configured slightly longer than the check_period.
